src/failures/src/communication/writers/transaction-writer.ts

```diff
@@ -0,0 +1,107 @@
+import { Reader, Updatable, Writer } from '..';
+
+/**
+ * Provides transactional writing on top of another writer.
+ *
+ * Requires that the written variable is read by a transactional reader.
+ */
+export class TransactionWriter implements Writer {
+    private retryAfterNumberOfUpdates: number;
+
+    private valueSimVar: Writer & Updatable;
+
+    private transactionSimVar: Reader & Writer;
+
+    private openTransactions: Record<number, Transaction> = {};
```

Comment:
This has hurt me in the past, because all Javascript object keys are strings internally. I would suggest either using a `Map`, or explicitly casting the keys to strings in your code to avoid future confusion about Javascript automatically casting the type:

```
> x[1] = "bar"
'bar'
> x
{ '1': 'bar' }
> Object.keys(x)
[ '1' ]
```